src/types/memoryview.rs

```diff
@@ -0,0 +1,123 @@
+use std::marker::PhantomData;
+use std::os::raw::c_char;
+
+use crate::{AsPyPointer, PyAny, PyResult, Python};
+use crate::buffer::{Element, PyBuffer};
+use crate::ffi::{self, Py_ssize_t};
+use crate::types::PySequence;
+
+/// Represents a Python `memoryview`.
+#[repr(transparent)]
+pub struct PyMemoryView<T>(PyAny, PhantomData<T>);
+
+pyobject_native_type_core!(PyMemoryView<T>, ffi::PyMemoryView_Type, #checkfunction=ffi::PyMemoryView_Check; T);
+
+impl PyMemoryView<u8> {
+    pub fn new<'p>(py: Python<'p>, memory: &mut [u8], is_writable: bool) -> &'p PyMemoryView<u8> {
```

Comment:
Unfortunately this API isn't safe - the memory would have to be borrowed for as long as the returned memoryview lives, but that can't be encoded since Python objects can't have associated lifetimes (except for the `'py` pseudo-lifetime).

It might be possible to offer it as `unsafe fn` with appropriate documentation.

Reply:
Agreed this isn't safe as-is. We technically already have the `unsafe fn` in `from_ptr`. It would also be _very_ hard to use correctly, as if you pass this object to other Python code it's extremely likely this memoryview may live longer than you expect.

A safe alternative may be to accept `&'static [u8]`, although that impairs usability a bit.

Maybe the `PyMemoryView::from` already started below is the thing we should encourage users to use (they can then wrap their buffers in `#[pyclass]` types which expose buffer access).

Reply:
Removed